File: trunk/Base Libraries/Base Render Library/Camera Control/Camera/Camera.cs
/*
 * Author: Bogolepov Denis (denisbogol@sandy.ru)
 */
 
using System;
using Base.MathTools;
using Tao.OpenGl;

namespace Base.Graphics
{
    [Serializable]
    public class Camera
    {
        #region Public Fields       
        
        public int Width = 512;

        public int Height = 512;

        public float FieldOfView = 45.0f;  

        #endregion
        
        #region Private Fields
        
        #region Положение и ориентация камеры

        private Vector3 position = new Vector3(20.0f, 0.0f, 0.0f);
        
        private Vector3 orientation = Vector3.Empty;

        #endregion        
        
        #region Стандартные направления камеры
        
        private Vector3 upDirection = Vector3.UnitZ;

        private Vector3 rightDirection = Vector3.UnitY;

        private Vector3 viewDirection = -Vector3.UnitX;

        #endregion
        
        #endregion

        #region Constructor

        public Camera() { }

        public Camera(Vector3 position)
        {
            Position = position;
        }

        public Camera(Vector3 position, Vector3 orientation)
            : this(position)
        {
            Orientation = orientation;
        }

        #endregion

        #region Private Methods

        private void UpdateDirections()
        {
            Matrix3 rotation = Matrix3.Rotate(orientation);

            viewDirection = -rotation * Vector3.UnitX;
            
            upDirection = rotation * Vector3.UnitZ;
            
            rightDirection = rotation * Vector3.UnitY;
        }

        #endregion

        #region Public Methods

        public void MoveStraight(float delta)
        {
            Position += viewDirection * delta;
        }

        public void MoveHorizontal(float delta)
        {
            Position += rightDirection * delta;
        }

        public void MoveVertical(float delta)
        {
            Position += upDirection * delta;
        }
        
        public void Rotate(Vector3 delta)
        {
            Orientation += delta;
        }
        
        public void Setup()
        {
            Gl.glMatrixMode(Gl.GL_PROJECTION); 
            
            Gl.glLoadIdentity(); 
            
            Glu.gluPerspective(FieldOfView, Aspect, 1.0f, 1000.0f);
            
         	Glu.gluLookAt(position.X, position.Y, position.Z,
                          position.X + ViewDirection.X,
                          position.Y + ViewDirection.Y,
                          position.Z + ViewDirection.Z,
                          UpDirection.X, UpDirection.Y, UpDirection.Z);
            
            Gl.glMatrixMode(Gl.GL_MODELVIEW);
        }
        
        public void SetShaderData(ShaderProgram program)
        {
        	program.SetUniformVector("Camera.Position", position);
        	
        	program.SetUniformVector("Camera.View", viewDirection);

        	program.SetUniformVector("Camera.Right", rightDirection * ScreenScale.X);
        	
        	program.SetUniformVector("Camera.Up", upDirection * ScreenScale.Y);        	        	
        	
        	program.SetUniformVector("Camera.PixelSize", PixelSize);
        }       
        
        #endregion

        #region Properties
        
        public float Aspect
        {
        	get
        	{
        		return Width / (float) Height;
        	}
        }

        public Vector2 ScreenScale
        {
        	get
        	{
        		return new Vector2((float) Math.Tan(Aspect * FieldOfView / 2.0), (float) Math.Tan(FieldOfView / 2.0));
        	}
        }
        
        public Vector2 PixelSize
        {
        	get        		
        	{
        		return new Vector2(1.0f / Width, 1.0f / Height);
        	}
        }        
        
        public Vector3 Position
        {
        	get
        	{
        		return position;
        	}
        	
        	set
        	{
        		position = value;
        	}
        }

        public Vector3 Orientation
        {
            get
            {
                return orientation;
            }
            
            set
            {
            	orientation = value;
            	
            	UpdateDirections();
            }
        }

        public Vector3 ViewDirection
        {
            get
            {
                return viewDirection;
            }
        }

        public Vector3 UpDirection
        {
            get
            {
                return upDirection;
            }
        }

        public Vector3 RightDirection
        {
            get
            {
                return rightDirection;
            }
        }
        
        #endregion
    }
}

File: trunk/Source/Base Libraries/Base Graphics Library/Camera.cpp
/*
   -----------------------------------------------------------------------------
   |                 B A S E   G R A P H I C S   L I B R A R Y                 |
   -----------------------------------------------------------------------------
                              
   Copyright (c) 2009 - 2010 Denis Bogolepov ( denisbogol @ gmail.com )

   This library is free software: you can redistribute it and/or modify it under
   the terms of the GNU General Public License as published by the Free Software
   Foundation, either version 3 of the License, or (at your option) any later
   version.

   This library is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
   FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along with
   this library. If not, see <http://www.gnu.org/licenses/>.
 */

#include "Camera.hpp"

namespace graphics
{
    /************************************************************************/
    /*                      CONSTRUCTOR AND DESTRUCTOR                      */
    /************************************************************************/

    Camera :: Camera ( Vector3f position,
                       Vector3f orientation )
    {
        fPosition = position;

        fOrientation = orientation;

        RotateAxes ( );
    }

    /************************************************************************/
    /*                           PRIVATE METHODS                            */
    /************************************************************************/

    void Camera :: RotateAxes ( void )
    {
        Matrix3f rotation;
        
        rotation =
            AngleAxisf ( fOrientation.z ( ), Vector3f :: UnitZ ( ) ) *
            AngleAxisf ( fOrientation.y ( ), Vector3f :: UnitY ( ) ) *
            AngleAxisf ( fOrientation.x ( ), Vector3f :: UnitX ( ) );

        //----------------------------------------------------------------

        fView = rotation * Vector3f :: UnitZ ( );

        fUp = rotation * Vector3f :: UnitY ( );

        fSide = rotation * Vector3f :: UnitX ( );

        //----------------------------------------------------------------

        fView.normalize ( );

        fUp.normalize ( );

        fSide.normalize ( );
    }

    /************************************************************************/
    /*                            PUBLIC METHODS                            */
    /************************************************************************/

    void Camera :: Setup ( void )
    {
        glMatrixMode ( GL_PROJECTION );

        glLoadIdentity ( );

        gluPerspective (
            fFieldOfView,
            fAspect,
            fNearPlane,
            fFarPlane );
        
        //-------------------------------------------------

        glMatrixMode ( GL_MODELVIEW );

        glLoadIdentity ( );

        gluLookAt (
            fPosition.x ( ),
            fPosition.y ( ),
            fPosition.z ( ),
            fPosition.x ( ) + fView.x ( ),
            fPosition.y ( ) + fView.y ( ),
            fPosition.z ( ) + fView.z ( ),
            fUp.x ( ),
            fUp.y ( ),
            fUp.z ( ) );
    }
    
    /************************************************************************/

    Vector3f Camera :: Position ( void ) const
    {
        return fPosition;
    }

    /************************************************************************/

    Vector3f Camera :: Orientation ( void ) const
    {
        return fOrientation;
    }

    /************************************************************************/

    Vector3f Camera :: View ( void ) const
    {
        return fView;
    }

    /************************************************************************/

    Vector3f Camera :: Up ( void ) const
    {
        return fUp;
    }

    /************************************************************************/

    Vector3f Camera :: Side ( void ) const
    {
        return fSide;
    }

    /************************************************************************/
    
    Vector2f Camera :: Scale ( void ) const
    {
        float tangent = tanf ( fFieldOfView / 2.0F );
        
        return Vector2f ( fAspect * tangent, tangent );
    }

    /************************************************************************/

    GLfloat Camera :: NearPlane ( void ) const
    {
        return fNearPlane;
    }

    /************************************************************************/

    GLfloat Camera :: FarPlane ( void ) const
    {
        return fFarPlane;
    }

    /************************************************************************/

    GLfloat Camera :: FieldOfView ( void ) const
    {
        return fFieldOfView;
    }

    /************************************************************************/

    GLuint Camera :: Width ( void ) const
    {
        return fWidth;
    }

    /************************************************************************/

    GLuint Camera :: Height ( void ) const
    {
        return fHeight;
    }

    /************************************************************************/

    GLuint Camera :: Left ( void ) const
    {
        return fLeft;
    }

    /************************************************************************/

    GLuint Camera :: Top ( void ) const
    {
        return fTop;
    }

    /************************************************************************/

    GLfloat Camera :: Aspect ( void ) const
    {
        return fAspect;
    }

    /************************************************************************/

    void Camera :: SetPosition ( Vector3f position )
    {
        fPosition = position;
    }

    /************************************************************************/

    void Camera :: SetOrientation ( Vector3f orientation )
    {
        fOrientation = orientation;

        RotateAxes ( );
    }

    /************************************************************************/

    void Camera :: SetViewport ( GLuint width,
                                 GLuint height,
                                 GLuint left,
                                 GLuint top )
    {
        fWidth = width;

        fHeight = height;

        fLeft = left;

        fTop = top;

        fAspect = width / ( GLfloat ) height;

        //-------------------------------------------------

        glViewport (
            fLeft,
            fTop,
            fWidth,
            fHeight );
    }

    /************************************************************************/

    void Camera :: SetViewFrustum ( GLfloat fieldOfView,
                                    GLfloat nearPlane,
                                    GLfloat farPlane )
    {
        fFieldOfView = fieldOfView;

        fNearPlane = nearPlane;

        fFarPlane = farPlane;
    }

    /************************************************************************/

    void Camera :: SetShaderData ( ShaderProgram * program )
    {
        program->SetUniformVector ( "Camera.Position", fPosition );

        program->SetUniformVector ( "Camera.Side", fSide );

        program->SetUniformVector ( "Camera.Up", fUp );

        program->SetUniformVector ( "Camera.View", fView );

        program->SetUniformVector ( "Camera.Scale", Scale ( ) );
    }
}